File: src/nautilus_ctp_adapter/runtime/session.py
from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum

from .models import CtpRuntimeCommand, CtpRuntimeCommandKind, CtpRuntimeEvent, CtpRuntimeEventKind


class CtpSessionState(StrEnum):
    DISCONNECTED = "disconnected"
    CONNECTING = "connecting"
    CONNECTED = "connected"
    AUTHENTICATED = "authenticated"
    LOGGED_IN = "logged_in"


@dataclass(slots=True)
class CtpSessionRuntime:
    state: CtpSessionState = CtpSessionState.DISCONNECTED

    def on_command(self, command: CtpRuntimeCommand) -> None:
        if command.kind is CtpRuntimeCommandKind.CONNECT:
            self.state = CtpSessionState.CONNECTING
        elif command.kind is CtpRuntimeCommandKind.DISCONNECT:
            self.state = CtpSessionState.DISCONNECTED

    def on_event(self, event: CtpRuntimeEvent) -> None:
        if event.kind is CtpRuntimeEventKind.CONNECTED:
            self.state = CtpSessionState.CONNECTED
        elif event.kind is CtpRuntimeEventKind.AUTH_SUCCEEDED:
            self.state = CtpSessionState.AUTHENTICATED
        elif event.kind is CtpRuntimeEventKind.LOGIN_SUCCEEDED:
            self.state = CtpSessionState.LOGGED_IN
        elif event.kind in {
            CtpRuntimeEventKind.DISCONNECTED,
            CtpRuntimeEventKind.AUTH_FAILED,
            CtpRuntimeEventKind.LOGIN_FAILED,
        }:
            self.state = CtpSessionState.DISCONNECTED

    @property
    def is_connected(self) -> bool:
        return self.state in {
            CtpSessionState.CONNECTED,
            CtpSessionState.AUTHENTICATED,
            CtpSessionState.LOGGED_IN,
        }

File: rust/ctp_runtime_core/src/session.rs
use crate::commands::{CtpCommand, CtpCommandKind};
use crate::events::{CtpEvent, CtpEventKind};

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub enum CtpSessionState {
    #[default]
    Disconnected,
    Connecting,
    Connected,
    Authenticated,
    LoggedIn,
}

#[derive(Debug, Clone, Default)]
pub struct CtpSessionRuntime {
    state: CtpSessionState,
}

impl CtpSessionRuntime {
    pub fn on_command(&mut self, command: &CtpCommand) {
        match command.kind {
            CtpCommandKind::Connect => self.state = CtpSessionState::Connecting,
            CtpCommandKind::Disconnect => self.state = CtpSessionState::Disconnected,
            _ => {}
        }
    }

    pub fn on_event(&mut self, event: &CtpEvent) {
        match event.kind {
            CtpEventKind::Connected => self.state = CtpSessionState::Connected,
            CtpEventKind::AuthSucceeded => self.state = CtpSessionState::Authenticated,
            CtpEventKind::LoginSucceeded => self.state = CtpSessionState::LoggedIn,
            CtpEventKind::Disconnected | CtpEventKind::AuthFailed | CtpEventKind::LoginFailed => {
                self.state = CtpSessionState::Disconnected
            }
            _ => {}
        }
    }

    pub fn state(&self) -> CtpSessionState {
        self.state
    }

    pub fn is_connected(&self) -> bool {
        matches!(
            self.state,
            CtpSessionState::Connected | CtpSessionState::Authenticated | CtpSessionState::LoggedIn
        )
    }
}
